File: cpp/0846-hand-of-straights.cpp
#include <map>
#include <vector>

class Solution {
    public:
        [[nodiscard]] auto isNStraightHand(
            const std::vector<int> &hand,
            const int group_size
        ) -> bool {
            auto counts = std::map<int, int>{};
            for (const auto card : hand) {
                ++counts[card];
            }

            for (const auto [start, start_freq] : counts) {
                if (start_freq > 0) {
                    for (auto i = start; i < start + group_size; ++i) {
                        counts[i] -= start_freq;
                        if (counts[i] < 0) {
                            return false;
                        }
                    }
                }
            }

            return true;
        }
};

File: py/0846-hand-of-straights.py
from collections import Counter


class Solution:
    def isNStraightHand(self, hand: list[int], group_size: int) -> bool:
        counts = Counter(hand)

        for start in sorted(counts):
            start_freq = counts[start]
            if start_freq > 0:
                for i in range(start, start + group_size):
                    counts[i] -= start_freq
                    if counts[i] < 0:
                        return False

        return True
